libraries/standard/mqtt/src/mqtt.c

```diff
@@ -375,41 +439,61 @@ MQTTStatus_t MQTT_Publish( MQTTContext_t * const pContext,
 
     if( status == MQTTSuccess )
     {
-        /* Send header first. */
-        bytesSent = sendPacket( pContext,
-                                pContext->networkBuffer.pBuffer,
-                                headerSize );
-
-        if( bytesSent < 0 )
+        /* Reserve state for publish message. Only to be done for QoS1 or QoS2. */
+        if( pPublishInfo->qos > MQTTQoS0 )
         {
-            LogError( "Transport send failed for PUBLISH header." );
-            status = MQTTSendFailed;
+            status = MQTT_ReserveState( pContext,
+                                        packetId,
+                                        pPublishInfo->qos );
         }
-        else
-        {
-            LogDebugWithArgs( "Sent %d bytes of PUBLISH header.",
-                              bytesSent );
+    }
 
-            /* Send Payload. */
-            bytesSent = sendPacket( pContext,
-                                    pPublishInfo->pPayload,
-                                    pPublishInfo->payloadLength );
+    if( status == MQTTSuccess )
+    {
+        /* Sends the serialized publish packet over network. */
+        status = sendPublish( pContext,
+                              pPublishInfo,
+                              headerSize );
 
-            if( bytesSent < 0 )
-            {
-                LogError( "Transport send failed for PUBLISH payload." );
-                status = MQTTSendFailed;
-            }
-            else
+        /* TODO. When a publish fails, the reserved state has to be cleaned
+         * up. This will have to be done once an API in state machine is
+         * available. */
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* Update state machine after PUBLISH is sent.
+         * Only to be done for QoS1 or QoS2. */
+        if( pPublishInfo->qos > MQTTQoS0 )
+        {
+            /* TODO MQTT_UpdateStatePublish will be updated to return
+             * MQTTStatus_t instead of MQTTPublishState_t. Update the
+             * code when that change is made. */
+            publishStatus = MQTT_UpdateStatePublish( pContext,
+                                                     packetId,
+                                                     MQTT_SEND,
+                                                     pPublishInfo->qos );
+
+            if( publishStatus == MQTTStateNull )
             {
-                LogDebugWithArgs( "Sent %d bytes of PUBLISH payload.",
-                                  bytesSent );
+                LogErrorWithArgs( "Update state for publish failed with status =%u."
+                                  " However PUBLISH packet is sent to the broker."
+                                  " Any further handling of ACKs for the packet Id"
+                                  " will fail.",
+                                  publishStatus );
```

Comment:
Don't really need this since we know it's MQTTStateNull, but we're going to change this on refactor anyway.